File: gpg-remailer/mailer/send.cc
#include "mailer.ih"

void Mailer::send(std::string contents, 
                  std::vector<std::string> const &recipients,
                  string subject)
{
    writeMailContents(d_mailName);

    string const &logLabel = label();

    for (auto &recipient: recipients)
        sendMail(mailCommand(recipient), logLabel, recipient, subject);
}

File: gpg-remailer/mailer/send.f
template <class Derived>
void Mailer<Derived>::send(std::string const &mailData, 
                       std::vector<std::string> const &configuredRecipients, 
                       bool dontSend)
{
    static_cast<Derived *>(this)->writeMailContents(mailData);

    std::vector<std::string> oneRecipient;

    std::vector<std::string> const &recipients = 
                setRecipients(oneRecipient, configuredRecipients);

    std::string logLabel = static_cast<Derived *>(this)->label();

    for (auto &recipient: recipients)
        sendMail(static_cast<Derived *>(this)->mailCommand(recipient), 
                 logLabel, recipient, dontSend);
}
